src, test: node internals' postmortem metadata

mmarchini opened the pull request:
Those changes are the first steps towards allowing debug tools to
navigate some of Node's internals strucutres, giving more possibilities
to developers doing post-mortem debugging. One example of what can be
achieved with the symbols added is a new command being developed for
llnode, which prints information about handles and requests on the
queue for a core dump file.

Ref: https://github.com/nodejs/post-mortem/issues/46

Obs.: To add good tests for those changes, something like nodejs/build#777 is needed.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
* Building system
* Tools

addaleax commented on src/env.cc:
```diff
@@ -23,6 +24,8 @@ using v8::Message;
 using v8::StackFrame;
 using v8::StackTrace;
 
+Environment *Environment::currentEnvironment;
```
style nit: align the `*` to the left (i.e. `Environment* Environment::currentEnvironment`)

addaleax commented on src/env.h:
```diff
@@ -360,6 +360,7 @@ class IsolateData {
 
 class Environment {
  public:
+  static Environment *currentEnvironment;
```
(ditto)

addaleax commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,140 @@
+        name="currentEnvironment",
+        value="(uint64_t) &Environment::currentEnvironment",
+        headers=["env.h"],
+        type_="uint64_t",
```
uintptr_t?

addaleax commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,140 @@
+        type_="uint64_t",
+    ),
+    DebugSymbol(
+        name="class__BaseObject__persistant_handle",
```
typo: `persistant`

addaleax commented on common.gypi:
```diff
@@ -50,6 +50,7 @@
       }, {
         'os_posix': 1,
         'v8_postmortem_support%': 'true',
+        'node_postmortem_support%': 'true',
```
What’s the motivation for making this optional?

addaleax commented on src/req-wrap.h:
```diff
@@ -19,6 +19,10 @@ class ReqWrap : public AsyncWrap {
   inline ~ReqWrap() override;
   inline void Dispatched();  // Call this after the req has been dispatched.
   T* req() { return &req_; }
+  #ifdef NODE_POSTMORTEM_SUPPORT
```
I don’t think these guards are necessary either way.

addaleax commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,140 @@
+    ),
+    DebugSymbol(
+        name="class__Environment__handleWrapQueue",
+        value="(size_t) (((Environment*)0)->handle_wrap_queue())",
```
Tbh, I’d just put `#define private public` at the top of the generated file and use `offsetof`.

bnoordhuis commented:
The changes to src/ seem rather ad hoc.  I'm also not a fan of adding global variables.  You don't need to because the env can be found by following `node_isolate`.

mmarchini commented:
Thanks for the feedbacks, I just updated the PR with the suggestions made by @addaleax. I'm just not sure if `#define private public` is safe enough to be used here (seems rather extreme to me), but it worked like a charm. 

All changes previously made on `src/` were removed (with the exception of `Environment::currentEnvironment`). Now I'll take a look into `node_isolate` to see how to access the environment from there (it's a little challenging because it's not possible to call functions or methods when debugging a core file, but I think it can be done). I'll update this PR after finding a better way to access the current environment.

mmarchini commented on node.gyp:
```diff
@@ -152,7 +152,8 @@
         'node_js2c#host',
-        'deps/nghttp2/nghttp2.gyp:nghttp2'
+        'deps/nghttp2/nghttp2.gyp:nghttp2',
+        'node-postmortem-metadata'
```
Is this the right place to include `node-postmortem-metadata` as a dependency?

bnoordhuis replied:
Seems fine to me but you should add a `#host` at the end.  For consistency, I'd name it `node_postmortem_metadata`, i.e., underscores instead of dashes.

mmarchini replied:
I'll make those changes. Should I also rename `gen-postmortem-metadata.py` to `gen_postmortem_metadata.py` (most files on `tools/` are using dashes, but there are some files using underscores)?

mmarchini commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+ */
+
+#define private public
+#define protected public
```
Is it safe to be used here?

bnoordhuis replied:
Seems fine to me.

rnchamberlain commented:
@bnoordhuis @mmarchini the isolate is a v8 class, the env is in node, and it looks to me that you could get the isolate from the env, but I don't see how you can go the other way?

mmarchini commented:
@rnchamberlain the Environment is stored in a fixed position inside the Context, and the Context is reachable from the Isolate, but the offsets to get the Context from the Isolate and the position inside the Context where the Environment is stored are currently not available in the symbols table of V8. I'm writing a PR to add those symbols to V8, let's see how it goes.

Another option would be to get the environment from `uv_key_t thread_local_env`, but it seems way harder to do, and any solution would be platform dependent (for example, on Linux the environment is stored in a pthread's thread-specific store).

mmarchini commented:
Pull Request updated: 
* Remove `currentEnvironment` symbol 
* Add `environment_context_idx_embedder_data` symbol (required to get the Environment from `node_isolate`)

Please let me know if there's any other changes needed.

BridgeAR commented:
@nodejs/build @nodejs/post-mortem PTAL

BridgeAR commented:
Again @nodejs/build @nodejs/post-mortem PTAL
Maybe @addaleax again as well?

addaleax commented:
> Maybe @addaleax again as well?

The code LGTM but I don’t think I have a sufficiently good feeling for the usefulness of this feature.

bnoordhuis commented on node.gyp:
```diff
@@ -717,7 +719,28 @@
+      'variables': {},
+      'actions': [
+        {
+          'action_name': 'gen-postmortem-metadata',
```
Can you add `'process_outputs_as_sources': 1,` here?

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python
```
Can you use two space indent in this file?

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+
+class DebugSymbol(object):
+    type_ = 'int'
+    _PREFIX = 'nodedbg_'
```
Inconsistent casing of class variables.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+        self.name = name
+        self.value = value
+        self.headers = headers
+        self.type_ = type_ or DebugSymbol.type_
```
Why is this called `type_` instead of `type`?

mmarchini replied:
There's a built-in function named `type`, so using it as a name for variables or attributes is usually discouraged, but I could rename that if you prefer.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+        headers = [debug_symbol.headers for debug_symbol in debug_symbols]
+        headers = sum(headers, [])
+
+        return [h for h in headers if not (h in seen or seen_add(h))]
```
This seems like a convoluted way of doing:
```python
headers = set()
for d in debug_symbols:
  headers.update(d.headers)
return list(headers)
```
Or even just:
```python
return list(set(sum(d.headers for d in debug_symbols, [])))
```

mmarchini replied:
I prefer those ways too, but the order the headers were declared should be preserved (some headers need to be included before others). I'll add a comment on this method explaining that.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+        return [h for h in headers if not (h in seen or seen_add(h))]
+
+    def __str__(self):
+        return "{type} {prefix}{name} = {value};".format(
```
Minor nit but we usually use single quotes for strings.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+    ),
+    DebugSymbol(
+        name="class__BaseObject__persistent_handle",
+        value="(size_t) offsetof(BaseObject, persistent_handle_)",
```
Unnecessary cast here and elsewhere: the return type of `offsetof()` is already `size_t`.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+    ))
+
+
+if (len(sys.argv) < 2):
```
Superfluous parens.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,146 @@
+
+
+if (len(sys.argv) < 2):
+    print('usage: %s output.cc' % sys.argv[0])
```
Can you use `.format()` here?

BridgeAR commented:
Ping @mmarchini

mmarchini commented:
Sorry for the late response. I've just updated the PR with changes requested by @bnoordhuis. I've also rebased it to match the current master.

@addaleax for now, it can be used to inspect active handles and requests on debuggers like LLDB (nodejs/llnode#122) and MDB (not implemented yet, but certainly will be if this is accepted). As we were discussing in nodejs/post-mortem#46, this is a first step to increase the power of debuggers to do post-mortem analysis of Node applications.

bnoordhuis commented on node.gyp:
```diff
@@ -169,7 +169,8 @@
         'node_js2c#host',
-        'deps/nghttp2/nghttp2.gyp:nghttp2'
+        'deps/nghttp2/nghttp2.gyp:nghttp2',
+        'node_postmortem_metadata#host'
```
If you swap these two, you can reduce the diff by one line (because you don't need to append a comma to the nghttp2 dep.)

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,149 @@
+  @classmethod
+  def get_headers(cls, debug_symbols):
+    '''
+    Return a list of headers without duplicates, preserving the order they were declared
```
Long line, can you wrap this at 80 columns?

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,149 @@
+    Return a list of headers without duplicates, preserving the order they were declared
+    '''
+    seen = set()
+    seen_add = seen.add  # faster than seen.add on each iteration
```
It probably makes no measurable difference.  I'd simply drop this and use `seen.add(h)` below.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,149 @@
+    headers = [debug_symbol.headers for debug_symbol in debug_symbols]
+    headers = sum(headers, [])
+
+    return [h for h in headers if not (h in seen or seen_add(h))]
```
`or seen_add(h)` is a somewhat misleading construct because `seen.add(h)` doesn't return anything (well, `None`.)

I'd spell it out for clarity:
```python
result = []
for h in headers:
  if not h in seen:
    seen.add(h)
    result.append(h)

return result
```

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,149 @@
+
+debug_symbols = [
+  DebugSymbol(
+    name="environment_context_idx_embedder_data",
```
Single quotes please.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,149 @@
+  out = file(sys.argv[1], 'w')
+  headers = DebugSymbol.get_headers(debug_symbols)
+  includes = ['#include "{}"'.format(header) for header in headers]
+  includes = "\n".join(includes)
```
Likewise.

mmarchini commented:
Pull Request updated with those last changes requested

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10210/

BridgeAR commented:
@mmarchini the commits should at least partially quashed and it is somewhat difficult for me to describe your changes in a single commit as I do not know what of the commit messages is still relevant and what not. Would you mind to either provide a single commit messages for all commits together or rebase on your own so this can be landed? Thanks a lot in advance!

mmarchini commented:
I've squashed all commits into one and updated the commit message to best reflect these changes. If you need anything else please let me know :)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10292/

BridgeAR commented:
Rerun the CI as it is very red https://ci.nodejs.org/job/node-test-commit/12650/

BridgeAR commented:
@mmarchini it seems like there are related failures on the CI. Please have a look. E.g. https://ci.nodejs.org/job/node-test-commit-linux/12698/nodes=centos5-64/console

mmarchini commented:
PR updated, there was a problem with string formatting when using Python 2.6. It should work now.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10368/

BridgeAR commented:
Rerun CI https://ci.nodejs.org/job/node-test-pull-request/10372/

BridgeAR commented:
I am not certain with this one but I have the feeling that there are more related failures on Windows.

mmarchini commented:
Apparently, VS2015 doesn't like `#define private public`, but there's a workaround for it ([_ALLOW_KEYWORD_MACROS](https://msdn.microsoft.com/en-us/library/bb531344(v=vs.140).aspx)). I updated the PR with this, but unfortunately, I don't have a Windows environment to test it (build ran normally on my Ubuntu 16.04). Can you run the CI again to see if Windows build is fixed now?

mmarchini commented:
Ping @BridgeAR

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11096/

gibfahn commented:
Seems there are still problems with windows: https://ci.nodejs.org/job/node-compile-windows/13032/label=win-vs2017/console

@mmarchini if you need access to a Windows box to debug we should be able to give you temporary access to one of ours (cc/ @nodejs/build)

Alternatively maybe one of the nice folks in @nodejs/platform-windows could have a look at this for you.

mmarchini commented:
@gibfahn I managed to get a Windows machine here at work, I'll try to build it here first. If I have a problem I'll contact the folks from @nodejs/platform-windows. Thanks

mmarchini commented:
Ok, I've found the problem. `#define private public` doesn't work on VisualStudio C++ because the visibility of class members is part of their name mangling schema, causing the linker to throw a lot of errors (see https://stackoverflow.com/a/23267346/2956796 and http://madebyevan.com/obscure-cpp-features/#redefining-keywords).

Trying more workarounds like this will probably cause some headache in the future, so I came up with a cleaner solution with no hacks or workarounds. This new approach uses a friend function to make private members accessible. I also added a macro (`ALLOW_DEBUG_SYMBOLS`), which makes a class become friend of that function.

Tested on Ubuntu 16.04, FreeBSD 11 and Windows 10 with Visual Studio 2017.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,169 @@
+  DebugSymbol(
+    name='class__BaseObject__persistent_handle',
+    value='offsetof(BaseObject, persistent_handle_)',
+    headers=['base-object.h', 'base-object-inl.h'],
```
You could drop `base-object.h` here.  We recently instituted a rule that the `*-inl.h` is sufficient if you want the implementation (as opposed to just the declaration, which is what the stem makes available.)

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,169 @@
+return 1;
+}}
+
+int debugSymbolsGenerated = genDebugSymbols();
```
Naming: `int debug_symbols_generated = GenDebugSymbol();`

As well, is it necessary to do it dynamically?  Doesn't `size_t nodedbg_foo_bar = offsetof(Foo, bar);` work?

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,169 @@
+    name='class__ReqWrapQueue__nextOffset',
+    value='offsetof(ListNode<ReqWrap<uv_req_t>>, next_)',
+    headers=['req-wrap.h', 'util.h'],
+    type_='size_t',
```
I don't remember if I mentioned it but I think you should be able to infer the type with `decltype()`.

mmarchini commented:
Ping @bnoordhuis

jasnell commented:
@mmarchini ... can I ask you to give this a quick rebase to fix the conflict in node.gyp, then I'll give it another CI run.

jasnell commented:
Assuming it passes the ci run, I'll get it landed :-)

mmarchini commented:
Rebased and tested on Ubuntu 16.04, FreeBSD 11 and Windows 10

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11585/

jasnell commented:
Failure in arm ci is unrelated.

jasnell commented:
Landed in 446c1ec

bnoordhuis commented:
See nodejs/node-v8#26, this breaks with V8 ToT.  Looks like the `get_standard_includes()` function is a tad too inclusive.

bnoordhuis commented:
It also generates a lot of `-Winvalid-offsetof` warnings...
```
/home/bnoordhuis/src/v1.x/out/Release/obj/gen/node-debug-support.cc:77:48: warning: offset of on non-standard-layout type 'node::BaseObject' [-Winvalid-offsetof]
nodedbg_class__BaseObject__persistent_handle = offsetof(BaseObject, persistent_handle_);
                                               ^                    ~~~~~~~~~~~~~~~~~~
# followed by umpteen similar warnings
```

joyeecheung commented:
I wonder does it actually work with `offsetof()`ing private members of abstract classes? I think this would need some surgery that somehow put those members into a POD to work properly? Is there an accompanying PR of debugger/plugins that uses those symbols?

mmarchini commented:
@bnoordhuis `get_standard_includes()` could have a static list of all C++ standard libraries instead of searching for them. What do you think?

@joyeecheung yes, nodejs/llnode#122 is working, but I agree that using `offsetof` on non-POD classes can lead to undefined behavior. I'll look into alternatives to `offsetof` before opening a new PR.

bnoordhuis commented:
Reopening due to revert in 088bba3.

bnoordhuis commented:
> get_standard_includes() could have a static list of all C++ standard libraries instead of searching for them. What do you think?

Sounds reasonable.

mmarchini commented:
PR updated with these changes:

* Statically added all C++ standard libraries used by Node and its dependencies (to avoid things like nodejs/node-v8#26)
* Added a macro to calculate the offset of members for non-POD classes

Regarding the `offsetof` problems, here is what I've found and what I did:

Some of those classes are POD, which means we can still use `offsetof` for them.

Two of the non-POD classes (`ReqWrap::req_wrap_queue_` and `HandleWrap::handle_wrap_queue_`) needs to have a member at a fixed offset because they are used by ContainerOf (see comment at `src/req_wrap.h:28`). Node does that by placing those members before any variable-size member. For these two we can use the same arithmetic used by `ContainerOf` to get the offset for the field. I wrote a macro for it:

```c++
#define offsetof_nonpod(CLASS, FIELD) reinterpret_cast<uintptr_t>\
        (&(static_cast<CLASS*>(0)->FIELD))
```

The other three (`Environment::req_wrap_queue_`, `Environment::handle_wrap_queue_` and `BaseObject::persistent_handle_`) respects the same constraint established in `src/req_wrap.h:28` (they don't have any member with variable size before the member we're calculating the offset), which means the same macro should work.

In short, that macro could be used for non-POD classes, as long as they follow the constraint described above. nodejs/llnode#122 keeps working with this.

@joyeecheung @bnoordhuis PTAL

mmarchini commented:
I'll try to write some tests for this feature tomorrow

joyeecheung commented:
@mmarchini I need to find some time to dig into this a bit more later. Meanwhile can you add a comment to those classes that need to follow the constraints so people don't accidentally break it?

mmarchini commented:
Update:
* Added comments to all relevant headers
* Added a page to docs explaining what those symbols are and where they are used (this page could be improved though)
* Added a few tests to validate the symbols generated

Regarding this last item, I was able to write 4 tests (out of 10). I'm having trouble to write a working test for `HandleWrap::handle_wrap_queue_`. My last attempt is throwing a segmentation fault:

```c++
TEST_F(DebugSymbolsTest, HandleWrapList) {
  v8::HandleScope handle_scope(isolate_);
  auto context = node::NewContext(isolate_);
  v8::Context::Scope context_scope(context);
  node::IsolateData* isolateData = node::CreateIsolateData(isolate_, uv_default_loop());
  Argv argv{"node", "-e", ";"};
  auto env = node::CreateEnvironment(isolateData, context, 1, *argv, 2, *argv);

  uv_handle_t handle_;


  auto obj_template = v8::FunctionTemplate::New(isolate_);
  obj_template->SetClassName(FIXED_ONE_BYTE_STRING(isolate_, "prop1"));
  obj_template->InstanceTemplate()->SetInternalFieldCount(1);

  v8::Local<v8::Object> object = obj_template->GetFunction()->NewInstance(context).ToLocalChecked();
  auto *obj = new TestHandleWrap(env, object, &handle_);
}
```

Any thoughts on how to fix or investigate it, or even another way to write those tests?

mmarchini commented on src/env.h:
```diff
@@ -721,8 +731,6 @@ class Environment {
 #endif
 
-  HandleWrapQueue handle_wrap_queue_;
-  ReqWrapQueue req_wrap_queue_;
```
For some reason `Environment::req_wrap_queue()` is completely broken in the cctest build (I tested it on the master branch too), it wasn't returning the right address for `req_wrap_queue_` (it was always off by -0x20 bytes), thus manipulating or accessing this queue was impossible on tests. Probably there's some flag or macro different from the binary build.

As there were no tests using `Environment::req_wrap_queue()` before it wasn't a problem, but with the introduction of `test_node_debug_support.cc` `Environment::req_wrap_queue()` should work, and moving this to the beginning of the class fixed it.

bnoordhuis replied:
Would it good to get to the bottom of that.  Moving data members around without really understanding the problem is not optimal.

mmarchini replied:
I spent a few days trying to get to the bottom of it, but without success. Best I can tell is that it looks like some alignment problem and/or some member which is conditionally compiled (either doesn't make complete sense, but was the best I could get after analyzing the problem for some time). I'll open an issue with what I was able to find, as this problem is also present on `master`

mmarchini replied:
Found the root of this problem. Solution proposed on: #17461

mhdawson commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+/* {{debug_symbols_file}} */
```
I like  test_node_postmortem_metadata.cc as at least for me it is more obvious what it is.

mmarchini replied:
Sounds clearer to me too, I'll change it

bnoordhuis commented on test/cctest/node_test_fixture.h:
```diff
@@ -109,4 +110,48 @@ class NodeTestFixture : public ::testing::Test {
+      environment_ = node::CreateEnvironment(isolate_data_,
+                                       context_,
+                                       1, *argv,
+                                       argv.nr_args(), *argv);
```
Argument alignment issues.

bnoordhuis commented on test/cctest/node_test_fixture.h:
```diff
@@ -109,4 +110,48 @@ class NodeTestFixture : public ::testing::Test {
+      return environment_;
+    }
+
+    v8::Local<v8::Context> context_;
```
Keep this private and add a getter.

bnoordhuis commented on test/cctest/node_test_fixture.h:
```diff
@@ -109,4 +110,48 @@ class NodeTestFixture : public ::testing::Test {
+    v8::Local<v8::Context> context_;
+
+   private:
+    v8::Context::Scope *context_scope_;
```
Star leans left (also in the other files - I won't point them out individually) but I'd just use a `std::unique_ptr`.

bnoordhuis commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+/*
+ * This file is a template used by ./tools/gen-postmortem-metadata.py to
+ * generate a proper test file
+ */
```
C++ comments?

bnoordhuis commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+   size_t self_size() const override { return sizeof(*this); }
+
+   TestHandleWrap(node::Environment* env, v8::Local<v8::Object> object, uv_handle_t* handle) :
+     node::HandleWrap(env, object, handle, node::AsyncWrap::PROVIDER_TIMERWRAP) {}
```
4 space indent.

bnoordhuis commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+   size_t self_size() const override { return sizeof(*this); }
+
+   TestReqWrap(node::Environment* env, v8::Local<v8::Object> object) :
+     node::ReqWrap<uv_req_t>(env, object, node::AsyncWrap::PROVIDER_TIMERWRAP) {}
```
Ditto.

bnoordhuis commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+};
+
+TEST_F(DebugSymbolsTest, ContextEmbedderDataIndex) {
+  EXPECT_EQ(nodedbg_environment_context_idx_embedder_data, node::Environment::kContextEmbedderDataIndex);
```
Long line.  Can you run `make lint`?  I'll stop pointing out style issues from here on.

mmarchini replied:
Sorry, I forgot to run `make lint` on my last commit. I'm running it and a spell-checker on comments right now, will update the PR soon.

mmarchini replied:
Out of curiosity, there are ~20 Python scripts on `tools/`, wouldn't it be good to have a Python linter as well? I can make a PR with this if there's interest in it

bnoordhuis commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+  Env env {handle_scope, isolate_, argv, this};
+
+  v8::Local<v8::Object> object = v8::Object::New(isolate_);
+  node::BaseObject *obj = new node::BaseObject(*env, object);
```
Leaks memory.  Use `std::unique_ptr`.

mmarchini replied:
I tried using `std::unique_ptr` and `delete`, but both are throwing an error at the end of the test because the referenced `v8::Object` still exists:

```
[----------] 6 tests from DebugSymbolsTest
[ RUN      ] DebugSymbolsTest.ContextEmbedderDataIndex
[       OK ] DebugSymbolsTest.ContextEmbedderDataIndex (4 ms)
[ RUN      ] DebugSymbolsTest.BaseObjectPersistentHandle
[18333]: ../src/base_object-inl.h:46:virtual node::BaseObject::~BaseObject(): Assertion `persistent_handle_.IsEmpty()' failed.
 1: node::Abort() [./out/Release/cctest]
 2: 0x14ef3ef [./out/Release/cctest]
 3: DebugSymbolsTest_BaseObjectPersistentHandle_Test::TestBody() [./out/Release/cctest]
 4: 0x128dabc [./out/Release/cctest]
 5: 0x128dd3d [./out/Release/cctest]
 6: 0x128df15 [./out/Release/cctest]
 7: 0x128f563 [./out/Release/cctest]
 8: testing::UnitTest::Run() [./out/Release/cctest]
 9: main [./out/Release/cctest]
10: __libc_start_main [/lib/x86_64-linux-gnu/libc.so.6]
11: _start [./out/Release/cctest]
Aborted (core dumped)
```

The failing check is:

```c++
inline BaseObject::~BaseObject() {
  CHECK(persistent_handle_.IsEmpty());
}
```

I'll try it again tomorrow, but suggestions are appreciated.

bnoordhuis replied:
Right, the destructor expects to be invoked by the garbage collector.  You can cheat and call `obj->persistent().Reset()` when you're done with it.

bnoordhuis commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+  node::BaseObject *obj = new node::BaseObject(*env, object);
+
+  auto expected = (uintptr_t)&(obj->persistent());
+  auto calculated = (uintptr_t)((void*)obj) + nodedbg_class__BaseObject__persistent_handle;
```
No C-style casts; use `static_cast` or `reinterpret_cast`.

bnoordhuis commented on test/cctest/test_node_debug_support.cc:
```diff
@@ -0,0 +1,127 @@
+  auto expected = (uintptr_t)obj;
+  auto calculated = next - nodedbg_class__ReqWrap__node;
+  EXPECT_EQ(expected, calculated);
+}
```
Please check with valgrind that the tests don't leak memory.

mmarchini replied:
Valgrind flagged one memory leak on `TEST_F(DebugSymbolsTest, BaseObjectPersistentHandle)`, which is the one you mentioned above.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,209 @@
+#
+# gen-postmortem-metadata.py output_file.cc
+#
+# Creates debugging symbols to help naviage Node's internals using post-mortem
```
navigate

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,209 @@
+
+
+debug_symbols = [
+  DebugSymbol(  # Done
```
Stray `# Done` comments here and below.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,209 @@
+    "iomanip", "iosfwd", "iostream", "istream", "iterator", "limits", "list",
+    "map", "memory", "new", "ostream", "queue", "set", "sstream", "stack",
+    "streambuf", "string", "tuple", "type_traits", "typeinfo", "unordered_map",
+    "unordered_set", "utility", "vector"]
```
Can you use single quotes?

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,209 @@
+  includes = '\n'.join(includes)
+
+  standard_includes = get_standard_includes()
+  standard_includes = ['#include <{0}>'.format(include) for include in standard_includes]
```
Long line.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,209 @@
+  out = file(sys.argv[2], 'w')
+  out.write(
+    test_template.replace('/* {{debug_symbols_file}} */', render_symbols_file())
+  )
```
Paren goes on previous line (which should be indented by 4 spaces and be <= 80 columns.)

bnoordhuis commented on node.gyp:
```diff
@@ -971,7 +975,30 @@
+          'action_name': 'gen-postmortem-metadata',
+          'process_outputs_as_sources': 1,
+          'inputs': [
+            './tools/gen-postmortem-metadata.py',
```
This is not enough to trigger a rebuild when object layouts change.  You need to take a dependency on the relevant headers in src/.

bnoordhuis commented on src/base_object.h:
```diff
@@ -65,6 +65,11 @@ class BaseObject {
+  // class because it is used by tools/gen-postmortem-metadata.py to calculate
+  // offsets and generate debug symbols for BaseObject, which assumes that the
+  // position of members in memory are predictable. Because of that, there must
+  // be no variable-size members before persistent_handle_
```
These comments are subtly wrong because there is no such thing as a variable-sized member.  What I suspect you're getting at is data members that are conditionally compiled in based on preprocessor defines.

mmarchini replied:
Variable-sized members can happen when a member type is based on the class template (which is not the case for BaseObject right now, but the comment is here to avoid problems if this changes in the future).

But I agree, this explanation should be improved. Maybe this could be explained in `doc/guides/node-postmortem-support.md` to avoid repetitive comments and to explain it better? Then this comment could be:

```c++
  // persistent_handle_ needs to be at a fixed offset from the start of the
  // class because it is used by tools/gen-postmortem-metadata.py to calculate
  // offsets and generate postmortem metadata for BaseObject. For more
  // information please refer to `doc/guides/node-postmortem-support.md`
```

What do you think?

bnoordhuis replied:
> Variable-sized members can happen when a member type is based on the class template

Okay, I understand what you mean now but I wouldn't use the word "variable-sized" for that, that's too easy to confuse with variable-length arrays and structs, but they are quite different beasts.

(And also because it's still subtly wrong: there's nothing variable-sized about them, `BaseObject<Foo>` and `BaseObject<Bar>` are simply different types.)

joyeecheung replied:
@bnoordhuis Maybe we can just leave it as `persistent_handle_ needs to be at a fixed offset from the start of the class.....which assumes the position of members in memory are predictable`, I think that is already clear enough.

davepacheco commented on doc/guides/debug-symbols.md:
```diff
@@ -0,0 +1,29 @@
+All symbols available for V8 are prefixed with `v8dbg_`, and they allow
+inspection of objects in the heap as well as those object's properties and
+references. Those symbols are provided by V8 itself by the script
+`deps/v8/tools/gen-postmortem-metadata.py`.
```
If this is intended for maintainers and we do want to refer to the script, you might want to mention that the script contains additional information about the various metadata that exists.

gibfahn replied:
If it's in `doc/guides`, then it should be for maintainers, user guides go in [nodejs/nodejs.org](https://github.com/nodejs/nodejs.org/).

mmarchini replied:
This document is intended for maintainers, but we could add a guide to nodejs/nodejs.org targeting users and tools developers. I'll update this document to make it clearer and more helpful.

mmarchini commented on doc/guides/debug-symbols.md:
```diff
@@ -0,0 +1,29 @@
+## Tools and References
+
+* [https://github.com/nodejs/llnode](llnode): LLDB plugin
+* [https://github.com/joynet/mdb_v8](mdb_v8): mdb plugin
```
Omg, I've been saying this wrong for some time :scream:

mmarchini commented:
Fixed styles and typos. I'll take a look at the other comments later this week.

mmarchini commented on test/cctest/test_node_postmortem_metadata.cc:
```diff
@@ -0,0 +1,135 @@
+  TestHandleWrap(node::Environment* env, v8::Local<v8::Object> object,
+                 uv_handle_t* handle) :
+      node::HandleWrap(env, object, handle, node::AsyncWrap::PROVIDER_TIMERWRAP)
+      {}
```
Is the indentation correct here?

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/11884/

joyeecheung commented:
Also can you rebase against master? `git am` does not apply cleanly with the patches.

joyeecheung commented:
Some failures look related to https://github.com/nodejs/node/issues/17043.

joyeecheung commented on doc/guides/node-postmortem-support.md:
```diff
@@ -0,0 +1,29 @@
+## Tools and References
+
+* [https://github.com/nodejs/llnode](llnode): LLDB plugin
+* [https://github.com/joynet/mdb_v8](mdb_v8): mdb plugin
```
Linter failed because `remark-lint-prohibited-strings` does not skip "v8" here, cc @Trott 

BTW links here are broken, I think you are going for things like `[llnode](https://github.com/nodejs/llnode)`?

Trott replied:
Link on line 27 should contain `joyent` rather than `joyent`.

Trott replied:
@joyeecheung The `mdb_v8` in the URL should be ignored by the linter. The `mdb_v8` meant for the link text will be fine if it is wrapped in backticks.

mmarchini commented:
Update:

* Fixed URLs on `doc/guides/node-postmortem-support.md`
* Rebased against master
* Reverted changes on `src/env.h`. Tests are broken because of that, but when https://github.com/nodejs/node/pull/17461 lands they'll work again
* Fixed style on `test/cctest/test_node_postmortem_metadata.cc` (I ran clang-format with settings from https://github.com/nodejs/node/pull/16122 on it)

Still need to:

* Rebase again when https://github.com/nodejs/node/pull/17461 lands
* Move explanation from comments to `doc/guides/node-postmortem-support.md`
* Maybe review symbol names? (@joyeecheung suggestions on that?)

It's probably best to wait before running CI on it again.

joyeecheung commented:
#17461 has landed. New CI: https://ci.nodejs.org/job/node-test-pull-request/12032/

joyeecheung commented:
Ooops, sorry, should have wait until this addresses the openssl conditional compilation issue...@mmarchini can you rebase again? Thanks

mmarchini commented:
Update:

* Rebased against master
* Updated comments on `src/` files to make them simpler
* Rewrite most of `doc/guides/node-postmortem-support.md`
* Changed all comments on new files to C++ comments (there were a couple of C comments left)

`make test` and `make lint` passed on Ubuntu 16.04. @joyeecheung can you rerun the CI, please?

MylesBorins commented:
I'm adding `dont-land` labels for all active release lines so that we don't accidentally re-add the reverted commit. Please open a backport to v9.x when this is ready to land.

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,208 @@
+
+// Need to import standard headers before redefining private, otherwise it
+// won't compile
+{standard_includes}
```
Since it's a static list now, can you write these out?

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,208 @@
+#define nonstandard_offsetof(CLASS, FIELD) reinterpret_cast<uintptr_t>\
+        (&(static_cast<CLASS*>(0)->FIELD))
+
+{includes}
```
This can be written out too, I think?

bnoordhuis commented on tools/gen-postmortem-metadata.py:
```diff
@@ -0,0 +1,208 @@
+
+rendered_symbols_file = render_symbols_file()
+create_symbols_file(rendered_symbols_file)
+create_test_file(rendered_symbols_file)
```
Correct me if I'm wrong but couldn't this script be replaced with a file in src/ now?  It doesn't do anything dynamic, it always produces the exact same output as far as I can tell.

bnoordhuis commented:
@mhdawson I didn't dismiss your review but you should check if this PR is still anything like what you reviewed.

mmarchini commented:
> Correct me if I'm wrong but couldn't this script be replaced with a file in src/ now? It doesn't do anything dynamic, it always produces the exact same output as far as I can tell.

I thought about that too, but then I tried to keep it similar to how V8 generates its symbols. The only thing done dynamically in the script right now is "injecting" `node_debug_support.cc` inside of `test_node_postmortem_metadata.cc`, as there's no header file to import the constants from, but I think this could be accomplished in another way.

I can create a file in `src/` instead, unless someone opposes to it and think it should stay as a script. Here's a Gist of how the file would look like https://gist.github.com/mmarchini/905f04115815105f8c86127ec165ad65

Any thoughts on this?

mmarchini commented:
@bnoordhuis I tried your patch and it worked :smile: 

I'm just not sure if there's an easier way to use it. Right now it looks like this (that's the worst case):

```c++
nodedbg_class__ReqWrapQueue__nextOffset =
    OffsetOf<ListNode<ReqWrap<uv_req_t>>*, ListNode<ReqWrap<uv_req_t>>>(
        &ListNode<ReqWrap<uv_req_t>>::next_);
```

I didn't update the Pull Request yet because I'm still making some improvements. Probably will update it on next Monday.

bnoordhuis commented:
The compiler should be able to infer the types.  Are you saying that doesn't work for you?  What compiler are you testing it with?

mmarchini commented:
I probably did something wrong the first time, because it wasn't inferring, but I tried again and it's working now.

```c++
nodedbg_class__ReqWrapQueue__nextOffset =
    OffsetOf(&ListNode<ReqWrap<uv_req_t>>::next_);
```

Way cleaner :)
Thanks!

mmarchini commented on test/cctest/test_node_postmortem_metadata.cc:
```diff
@@ -0,0 +1,138 @@
+#include "tracing/agent.h"
+#include "v8.h"
+
+const int node::Environment::kContextEmbedderDataIndex;
```
Without this line, the Debug build breaks, `make -j4 BUILDTYPE=Debug` throws the following error:

```
/home/matheus/workspace/sthima/node/out/Debug/obj.target/cctest/test/cctest/test_node_postmortem_metadata.o: In function `DebugSymbolsTest_ContextEmbedderDataIndex_Test::TestBody()':
/home/matheus/workspace/sthima/node/out/../test/cctest/test_node_postmortem_metadata.cc:46: undefined reference to `node::Environment::kContextEmbedderDataIndex'
collect2: error: ld returned 1 exit status
cctest.target.mk:217: recipe for target '/home/matheus/workspace/sthima/node/out/Debug/cctest' failed
make[1]: *** [/home/matheus/workspace/sthima/node/out/Debug/cctest] Error 1
```

Which is weird, because it works on `node_postmortem_metadata.cc`. Suggestions?

mmarchini replied:
Neither of them worked :(
Debug build still gives the same error

joyeecheung replied:
I guess the linker is complaining because`kContextEmbedderDataIndex` is ODR-used below, you can make it not ODR-used by applying lvalue-to-rvalue conversion immediately on it (e.g. assigning it to a `int` before using it in `EXPECT_EQ`).

mmarchini replied:
Worked, thanks! :tada:

```c++
TEST_F(DebugSymbolsTest, ContextEmbedderDataIndex) {
  int kContextEmbedderDataIndex = node::Environment::kContextEmbedderDataIndex;
  EXPECT_EQ(nodedbg_const_Environment__kContextEmbedderDataIndex__int,
            kContextEmbedderDataIndex);
}
```

mmarchini commented:
Huge update:

* Replaced `tools/gen-postmortem-metadata.py` with `src/node_postmortem_metadata.cc`
* Renamed all constants to more meaningful names (see below)
* Rebased against master
* Fixed styling problems and typos
* Fixed memory leak on cctests
* Updated `doc/guides/node-postmortem-support.md` and comments on `src/` files to match the new structure
* Updated `node.gyp` to match the new structure


`make test`, `make lint`, and `valgrind ./out/Release/cctest` green on Ubuntu 16.04.

Regarding the constants' names, I'm proposing the following format for them (as described in `src/node_postmortem_metadata.cc`):

```
nodedbg_const_CLASS__MEMBER      Describes a class member constant
nodedbg_offset_CLASS__MEMBER     Describes the offset to a class member
```

And the new names are:

```c++
// BaseObject::persistent_handle_ offset
uintptr_t nodedbg_offset_BaseObject__persistent_handle_;

// Environment::kContextEmbedderDataIndex value;
int nodedbg_const_Environment__kContextEmbedderDataIndex;
// Environment::handle_wrap_queue_ offset
uintptr_t nodedbg_offset_Environment__handle_wrap_queue_;
// Environment::req_wrap_queue_ offset
uintptr_t nodedbg_offset_Environment__req_wrap_queue_;

// HandleWrap::handle_wrap_queue_ offset
uintptr_t nodedbg_offset_HandleWrap__handle_wrap_queue_;
// HandleWrapQueue::head_ offset
uintptr_t nodedbg_offset_HandleWrapQueue__head_;
// ListNode<HandleWrap>::next_ offset
uintptr_t nodedbg_offset_ListNode_HandleWrap__next_;

// ReqWrap::req_wrap_queue_ offset
uintptr_t nodedbg_offset_ReqWrap__req_wrap_queue_;
// ReqWrapQueue::head_ offset
uintptr_t nodedbg_offset_ReqWrapQueue__head_;
// ListNode<ReqWrap>::next_ offset
uintptr_t nodedbg_offset_ListNode_ReqWrap__next_;
```

What do you think?

bnoordhuis commented on test/cctest/test_node_postmortem_metadata.cc:
```diff
@@ -0,0 +1,135 @@
+  const Argv argv;
+  Env env{handle_scope, argv, this};
+
+  uv_handle_t handle;
```
Can you use an initialized handle, like a `uv_tcp_t`?  Passing an uninitialized handle to HandleWrap may work now but that's not something I want to debug in the future.

bnoordhuis commented on test/cctest/test_node_postmortem_metadata.cc:
```diff
@@ -0,0 +1,135 @@
+
+  v8::Local<v8::Object> object =
+      obj_template->GetFunction()->NewInstance(env.context()).ToLocalChecked();
+  auto* obj = new TestHandleWrap(*env, object, &handle);
```
`obj` is leaked.

bnoordhuis commented on test/cctest/test_node_postmortem_metadata.cc:
```diff
@@ -0,0 +1,135 @@
+
+  v8::Local<v8::Object> object =
+      obj_template->GetFunction()->NewInstance(env.context()).ToLocalChecked();
+  auto* obj = new TestReqWrap(*env, object);
```
`obj` is leaked.

joyeecheung commented:
The new format LGTM, although I have a suggestion: can you add the type to the end (`nodedbg_const_CLASS__MEMBER__TYPE` and `nodedbg_offset_CLASS__MEMBER__TYPE`) like what V8 does for the similar constants? So it would be something like `nodedbg_offset_BaseObject__persistent_handle___v8_Persistent_v8_Object`, `nodedbg_const_Environment__kContextEmbedderDataIndex__int`, .etc

mmarchini commented:
@joyeecheung makes sense and certainly helps to avoid tools misinterpreting the symbols. I'll update the names with this format.

mmarchini commented:
Update:

* Renamed symbols with @joyeecheung suggestion
* Fixed memory leaks on `TEST_F(DebugSymbolsTest, HandleWrapList)` and 
  `TEST_F(DebugSymbolsTest, ReqWrapList)`
* Use `extern "C" { ... }` and `NODE_EXTERN` on 
  `node_postmortem_metadata.cc`
* Use `uv_tcp_t` instead of `uv_handle_t`
* Use macros to enforce symbols' name format
* Use macro expansion to declare symbols (something I decided to try, but we can revert it to the normal declaration if the code gets difficult to read/maintain)

P.S.: I didn't amend this commit to make it easier to diff all changes. Once we agree on the naming convention and the code structure, I'll squash them.

bnoordhuis commented on src/node_postmortem_metadata.cc:
```diff
@@ -58,54 +58,78 @@ int GenDebugSymbol();
-//
+#define NODEDBG_SYMBOL(Name)  nodedbg_ ## Name
+
+// nodedbg_const_CLASS__MEMBER__TYPE: Describes a class member constant
```
Dot at end.

bnoordhuis commented on src/node_postmortem_metadata.cc:
```diff
@@ -58,54 +58,78 @@ int GenDebugSymbol();
+#define NODEDBG_CONST(Class, Member, Type) \
+    NODEDBG_SYMBOL(const_ ## Class ## __ ## Member ## __ ## Type)
+
+// nodedbg_offset_CLASS__MEMBER__TYPE: Describes the offset to a class member
```
Ditto.

bnoordhuis commented on src/node_postmortem_metadata.cc:
```diff
@@ -58,54 +58,78 @@ int GenDebugSymbol();
-uintptr_t nodedbg_offset_ReqWrapQueue__head_;
-uintptr_t nodedbg_offset_ListNode_ReqWrap__next_;
+extern "C" {
+  #define V(Class, Member, Type)                                              \
```
Don't indent preprocessor directives.

bnoordhuis commented on src/node_postmortem_metadata.cc:
```diff
@@ -58,54 +58,78 @@ int GenDebugSymbol();
-uintptr_t nodedbg_offset_Environment__handle_wrap_queue_;
-uintptr_t nodedbg_offset_Environment__req_wrap_queue_;
+#define NODE_CONST_POSTMORTEM_METADATA(V)                                     \
+  V(Environment, kContextEmbedderDataIndex, int)
```
If there's only one, I'd write it out in full - that's going to be shorter and clearer than doing it through x-macros.

bnoordhuis commented on src/node_postmortem_metadata.cc:
```diff
@@ -58,54 +58,78 @@ int GenDebugSymbol();
       OffsetOf(&Environment::ReqWrapQueue::head_);
-  nodedbg_offset_ListNode_ReqWrap__next_ =
+  NODEDBG_OFFSET(ListNode_ReqWrap, next_, ListNode_ReqWrapQueue) =
       OffsetOf(&ListNode<ReqWrap<uv_req_t>>::next_);
```
Can you generate these with another x-macro?

bnoordhuis commented on test/cctest/node_test_fixture.h:
```diff
@@ -76,6 +76,13 @@ class NodeTestFixture : public ::testing::Test {
+
+    // As the TracingController is stored globally, we only need to create it
+    // one time for all tests.
+    if(node::tracing::TraceEventHelper::GetTracingController() == nullptr) {
```
Space before `(`.  Didn't the linter complain?

bnoordhuis commented on test/cctest/node_test_fixture.h:
```diff
@@ -76,6 +76,13 @@ class NodeTestFixture : public ::testing::Test {
+    // one time for all tests.
+    if(node::tracing::TraceEventHelper::GetTracingController() == nullptr) {
+      node::tracing::TraceEventHelper::SetTracingController(
+        new v8::TracingController());
```
4 space indent.

bnoordhuis commented on test/cctest/test_node_postmortem_metadata.cc:
```diff
@@ -53,7 +53,7 @@ TEST_F(DebugSymbolsTest, BaseObjectPersistentHandle) {
   auto expected = reinterpret_cast<uintptr_t>(&obj.persistent());
   auto calculated = reinterpret_cast<uintptr_t>(&obj) +
-      nodedbg_offset_BaseObject__persistent_handle_;
+      NODEDBG_OFFSET(BaseObject, persistent_handle_, v8_Persistent_v8_Object);
```
I think it will help legibility when the variable name is written out rather than going through a macro.  It's not wrong as such, just harder to parse for humans.

bnoordhuis commented on test/cctest/test_node_postmortem_metadata.cc:
```diff
@@ -87,49 +88,53 @@ TEST_F(DebugSymbolsTest, HandleWrapList) {
   auto queue = reinterpret_cast<uintptr_t>((*env)->handle_wrap_queue());
-  auto head = queue + nodedbg_offset_HandleWrapQueue__head_;
-  auto next = head + nodedbg_offset_ListNode_HandleWrap__next_;
+  auto head = queue +
```
Tiniest of nits but can you break after the `=` instead of the `+`?  (Also below.)

mmarchini commented:
Update:

* Dot at the end of comments
* Add new symbol for `NODE_EXTSTR_OFF_DATA` (requested by @davepacheco on nodejs/post-mortem#46)
* Generate all symbols with x-macro (there are two const symbols now, and probably more will come in the future)
* Fix style issues and nits

@bnoordhuis can you take a look at the new x-macros? If everything is ok, I'll squash the commits.


> I think it will help legibility when the variable name is written out rather than going through a macro. It's not wrong as such, just harder to parse for humans.

I agree, but sometimes the name can get big enough to make it unfeasible to fit in 80 characters. For example:

```c++
TEST_F(DebugSymbolsTest, EnvironmentHandleWrapQueue) {
  // Will not fit in 80 characters
  auto calculated = reinterpret_cast<uintptr_t>(*env) +
      nodedbg_offset_Environment__handle_wrap_queue___Environment_HandleWrapQueue;

  // Macros let us break lines whenever necessary
  auto calculated = reinterpret_cast<uintptr_t>(*env) +
      NODEDBG_OFFSET(Environment, handle_wrap_queue_,
                     Environment_HandleWrapQueue);
}
```

We could use the macro only in this case, but it would be weird and confusing for readers.

bnoordhuis commented:
> I agree, but sometimes the name can get big enough to make it unfeasible to fit in 80 characters.

Not a problem, you can silence the linter by tacking on a `// NOLINT(whitespace/line_length)` comment.  Although I'm pretty religious about long lines I don't have a problem with it here.

bnoordhuis commented on src/node_postmortem_metadata.cc:
```diff
@@ -76,61 +77,55 @@ int GenDebugSymbol();
-  V(Environment, kContextEmbedderDataIndex, int)
+  V(Environment__kContextEmbedderDataIndex, int,                              \
+    Environment::kContextEmbedderDataIndex)                                   \
+  V(kNodeOffExtstrData, size_t, NODE_OFF_EXTSTR_DATA)
```
Terrible name, `NODE_OFF_EXTSTR_DATA`... I admit I also don't really understand what purpose it serves since it's always the size of a pointer.

mmarchini commented:
Wrote the names out in the tests instead of using macros and squashed all commits into one.

> Terrible name, NODE_OFF_EXTSTR_DATA... I admit I also don't really understand what purpose it serves since it's always the size of a pointer.

Should I do something about it in this PR?

bnoordhuis commented:
I don't like adding things the purpose is not understood of so I'd leave it out until it's established it's actually useful for something.  Maybe @davepacheco can shed light on that.  LGTM apart from that.

mmarchini commented:
I removed `NODE_OFF_EXTSTR_DATA` for now, we can add it again in the future if deemed necessary.

davepacheco commented:
NODE_OFF_EXTSTR_DATA is the offset of the "data" field in Node's "ExternString" class.  It's used in both mdb_v8 and the V8 ustack helper (via src/v8abbr.h).

mmarchini commented:
@davepacheco thanks for the explanation. @bnoordhuis do you think it's enough to add back this constant to our postmortem metadata?

mmarchini commented:
nits addressed :)

bnoordhuis commented:
> do you think it's enough to add back this constant to our postmortem metadata?

Yep, but let's give it a better name.

mmarchini commented:
A few suggestions:

1. `EXTSTR_DATA_FIELD_OFFSET`
1. `NODE_EXTSTR_DATA_FIELD_OFFSET`
1. `EXTERN_STRING_DATA_OFFSET`
1. `EXTERN_STRING_DATA_FIELD_OFFSET`

I prefer `EXTERN_STRING_DATA_FIELD_OFFSET`, because it makes it clear what the constant is. The postmortem metadata could be `nodedbg_const_EXTERN_STRING_DATA_FIELD_OFFSET__int`. Any other ideas?

bnoordhuis commented:
Its V8 counterpart is called `v8dbg_class_ExternalString__resource__Object`, perhaps follow that convention if feasible and reasonable.

mmarchini commented:
What about `nodedbg_offset_ExternalString__Data__int`?

joyeecheung commented on src/node_postmortem_metadata.cc:
```diff
@@ -0,0 +1,117 @@
+    HandleWrap::handle_wrap_queue_)                                           \
+  V(Environment_HandleWrapQueue, head_, ListNode_HandleWrap,                  \
+    Environment::HandleWrapQueue::head_)                                      \
+  V(ListNode_HandleWrap, next_, ListNode_HandleWrap,                          \
```
Looking more carefully, the type of this should be a `uintptr_t`

joyeecheung commented on src/node_postmortem_metadata.cc:
```diff
@@ -0,0 +1,117 @@
+    ReqWrap<uv_req_t>::req_wrap_queue_)                                       \
+  V(Environment_ReqWrapQueue, head_, ListNode_ReqWrapQueue,                   \
+    Environment::ReqWrapQueue::head_)                                         \
+  V(ListNode_ReqWrap, next_, ListNode_ReqWrapQueue,                           \
```
This should be a `uintptr_t` as well.

joyeecheung commented:
I think following the current convention, the symbol should be `nodedbg_offset_ExternalString__data__uintptr_t`?

mmarchini commented:
PR updated with @joyeecheung suggestions and branch rebased again.

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12501/

joyeecheung commented:
Relavent failures on ARM:

```
../test/cctest/test_node_postmortem_metadata.cc:140: Failure
Value of: calculated
  Actual: 164437576
Expected: expected
Which is: 4398029288112
[  FAILED  ] DebugSymbolsTest.ReqWrapList (8 ms)
```

mmarchini commented:
I was able to replicate the failing test on a [Scaleway](https://www.scaleway.com/) ARM64 CentOS server, but the test works if I run with the Debug build or if I try to `std::cout` in some parts of the code. For example, if I `std::cout` before the `PushBack()` in the `ReqWrap` constructor, the test fails, but if I `std::cout` right after, the test works:

```c++
template <typename T>
ReqWrap<T>::ReqWrap(Environment* env,
                    v8::Local<v8::Object> object,
                    AsyncWrap::ProviderType provider)
    : AsyncWrap(env, object, provider) {

  // FIXME(bnoordhuis) The fact that a reinterpret_cast is needed is
  // arguably a good indicator that there should be more than one queue.
  std::cout << "Before PushBack" << std::endl;
  env->req_wrap_queue()->PushBack(reinterpret_cast<ReqWrap<uv_req_t>*>(this));
  // If I uncomment the next line the test passes
  // std::cout << "After PushBack" << std::endl;
}
```

Printing the queue reveals that when the test fails, the queue is empty as if `PushBack()` was not called. I think this might be some optimization done by `gcc`.

I also noticed that the `gcc` version used in the CI is not supported by Node, and the version I was using is the same from the CI:

```
WARNING: C++ compiler too old, need g++ 4.9.4 or clang++ 3.4.2 (CXX=g++)
```

Updating here to gcc 5.5 fixed the problem :thinking: 
I'll try to downgrade to 4.9.4 to see if the problem persists

joyeecheung commented:
> I also noticed that the gcc version used in the CI is not supported by Node

^ cc @nodejs/build any context on that?

rvagg commented:
> > I also noticed that the gcc version used in the CI is not supported by Node

> ^ cc @nodejs/build any context on that?

yeah, see this search https://github.com/nodejs/build/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+armv7

https://github.com/nodejs/build/issues/829 probably has the most context, getting a newer GCC on Wheezy is ... basically impossible, at least I haven't found a way to do it. V8's minimum requirements continue to creep up, we continue to bump up our "minimum compiler version" but our infra isn't keeping up with that, partly because we're producing binaries on certain Linuxes (libc versions) that we have to keep on producing for the life of an LTS branch.

For Node 10 we're going to need to make a call on this which will probably mean decisively cutting off Wheezy, Ubuntu 12.04 and possibly even RHEL 6 which is the same era but is supposed to be supported till the end of 2020 (and 2024 in extended!). So that's fun .. It'd be nice if we could slow down on this compiler creep though.

rvagg commented:
Policy discussion can happen here:  https://github.com/nodejs/build/issues/1087

In the meantime, you're just going to have to work with what we have unfortunately.

joyeecheung commented:
@mmarchini For this PR to proceed I think either we will have to find a way to get around the gcc bug, or we should skip the test on ARM hosts (not quite sure how to do it for cctest?)

mmarchini commented:
@joyeecheung can we run the CI again to see if this workaround works?

mmarchini commented:
Rebased fixing the conflict

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12634/

joyeecheung commented:
CI was happy, I think this should be ready to land.

Pinging people previously signed-off on this: @bnoordhuis @mcollina @mhdawson any thoughts before the it lands?

CI before landing: https://ci.nodejs.org/job/node-test-commit/15669/

joyeecheung commented:
Oh no, looks like the V8 v6.4 update https://github.com/nodejs/node/pull/17489 broke this PR...@mmarchini do you mind taking another look?

mmarchini commented:
No problem, I was already rebasing it locally to see what happened.

mmarchini commented:
Same error is happening on 990959d2f6bbcb42a437381bb2f435ec5d8067e6 (upstream master), and apparently with others (https://github.com/nodejs/node/pull/17489#issuecomment-360312521). Maybe it's unrelated?

joyeecheung commented:
@mmarchini Yes, we should wait until the CI gets back to normal, fix is opened in https://github.com/nodejs/node/pull/18360

mmarchini commented:
Rebased and tested after #18360 landed

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-commit/15679/
V8 CI: https://ci.nodejs.org/job/node-test-commit-v8-linux/1182/

mmarchini commented:
`parallel/test-tls-server-verify` failure on Windows is unrelated (https://github.com/nodejs/node/issues/18269)

`message/if-error-has-good-stack` failure on Windows and the failure on arm64_odroid_c2 also seems unrelated, although I'm not sure

joyeecheung commented:
`message/if-error-has-good-stack` is unrelated: https://github.com/nodejs/node/issues/18377

let's try to see if we can get arm64_odroid_c2 running again, there have been some infra issues on that one but I would like to make sure the test passes there since we are dealing with compiler quirks in this PR

joyeecheung commented:
New ARM CI: https://ci.nodejs.org/job/node-test-commit-arm/13367/

If we cannot get a pass on arm64_odroid_c2 by tomorrow, I suggest we land this first. We could always revert if it turns out this breaks arm64_odroid_c2, which is probably unlikely since we have passed there before the v8 6.4 update.

mmarchini commented:
Still failing, although it's another error now (same error from https://ci.nodejs.org/job/node-test-pull-request/12734/). Maybe the machine is running out of memory? I'm willing to help investigate the problem if help is needed

joyeecheung commented:
@mmarchini That's a known issue https://github.com/nodejs/build/issues/884

joyeecheung commented:
Another one...:https://ci.nodejs.org/job/node-test-commit-arm/13371/

joyeecheung commented:
CI before landing: https://ci.nodejs.org/job/node-test-pull-request/12738/

joyeecheung commented:
Landed in 756a34e86386bbc879234c4987eacdcb21e0e54b, thanks for being so patient and following along! @mmarchini

MSLaguana commented:
I just ran into an issue caused by the redefinition of `private` in this change: it means that any attempt at private inheritance (in the included files) are broken and made invalid. Is this an intentional trade-off, or something that was unanticipated?

joyeecheung commented:
@MSLaguana How would they be broken? Do you mean that their members are now all accessible to `GenDebugSymbols()`?

joyeecheung commented:
@bnoordhuis Just curious, is that a convention? There is `class StringSearch : private StringSearchBase` and `class Reference : private Finalizer` in our codebase. If that needs to be enforced, we should put that in the C++ style guide.

bnoordhuis commented:
I saw that by pure coincidence ten minutes after I posted that... the reason we don't use non-public inheritance is that there is never a reason to.  The two files you mention both came from outside.

Yes, a note in the style guide would be good.

MylesBorins commented:
How did this end up with the `dont-land-on-9.x` label?

That is in conflict with the open backport PR https://github.com/nodejs/node/pull/18550

It is also worth mentioning that the original commit https://github.com/nodejs/node/commit/756a34e86386bbc879234c4987eacdcb21e0e54b lands cleanly on v9.x-staging, so unsure why we need to PR to begin with

MylesBorins commented:
One more thing... should this be semver-minor?

mmarchini commented:
> How did this end up with the dont-land-on-9.x label?

AFAIK it's because there was one reverted commit here, see https://github.com/nodejs/node/pull/14901#issuecomment-350980223 (this comment also explains why I opened a backport PR).

mmarchini commented:
> One more thing... should this be semver-minor?

Yep, since it's a new feature.

MylesBorins commented:
@mmarchini thanks for pointing it out! We'll get it in the next release!

MylesBorins commented:
Do we want to backport this to v8.x???
